### DISCUSSION TITLE
Proposal: Calculate hash value for files in storage

### DISCUSSION THREAD

**Discussion by pktiuk** [open-webui/open-webui#7918] [Ideas]
I think it would be very helpful to add file hash to the database and allow easy searching for files by these file hashes.
(It would make syncing and updating files much easier for some clients using REST API).

proposed change for `File` class:

```diff
diff --git a/backend/open_webui/models/files.py b/backend/open_webui/models/files.py
index 4050b0140..f8df91e54 100644
--- a/backend/open_webui/models/files.py
+++ b/backend/open_webui/models/files.py
@@ -19,7 +19,8 @@ class File(Base):
     __tablename__ = "file"
     id = Column(String, primary_key=True)
     user_id = Column(String)
-    hash = Column(Text, nullable=True)
+    hash_content = Column(Text, nullable=True, comment="Hash (sha256) from content string")
+    hash_file = Column(Text, nullable=True, comment="Hash calculated from the file itself")
 
     filename = Column(Text)
```

**Comment by pktiuk** [comment]
Could I prepare PR with these changes?

**Reply by pktiuk** [reply]
@tjbck Are you interested in this feature at all?

**Comment by stoerr** [comment]
I'd like to support that. I was just writing a sync program https://github.com/stoerr/openwebui-knowledgesync that syncs from filesystem to knowledge base as well, since something like that is desperately needed, but have trouble since the sha256 hashes calculated on the files in many cases don't match the hashes that are saved in the database. Therefore it's quite difficult to find out whether to update a file.

By the way: why is the saved hash different from the hash of the binary file? At least in case of ASCII files it is identical, but in other cases sometimes not. If I could somehow reproduce the hash calculation outside, that would be helpful as well, though I'm doing it in Javascript.
